### PR TITLE
add conformance gate prow job

### PR DESCRIPTION
**Pull request by zachmandeville**
## Summary
This PR introduces a job to check whether any stable endpoints promoted in the release are without conformance tests.  
## Background
It is [a requirement that GA/stable API's have conformance tests](https://github.com/kubernetes/community/pull/1806).  This requirement is hard to enforce and multiple releases have had stable endpoints promoted without tests.  This creates technical debt for conformance,with test writers working on new tests to target API's from older releases.  You can see this in this graph from [APISnoop](https://apisnoop.cncf.io/conformance-progress):
visualization(3)
where the orange bars in each release mark newly promoted endpoints without tests, and the accumulation of technical debt.
## This job
This prow job uses the same [APISnoop database](https://github.com/cncf/apisnoop/tree/master/apps/snoopdb) to pull in the audit logs from the most recent runs of these two e2e jobs:
- [ci-kubernetes-e2e-gci-gce](https://testgrid.k8s.io/google-gce#gce-cos-master-default)
- [ci-kubernetes-gce-conformance-latest](https://testgrid.k8s.io/conformance-all#Conformance%20-%20GCE%20-%20master)
It compares the endpoints hit by conformance tests in the audit logs to the  endpoints defined in the open API spec, and lists any new GA endpoints that are not hit by a conformance test.  If the list is empty, the job succeeds.  If not, the job fails.  One can then look at the bottom of the logs to find a list of untested endpoints.
You can see an example of this job on [prow.cncf.io](https://prow.cncf.io).

This job is intended to ultimately work as a release blocking job, and so its configured to fit  the [release blocking criteria](https://github.com/kubernetes/sig-release/blob/master/release-blocking-jobs.md#release-blocking-criteria-and-dashboard).

### CODE REVIEW

**Comment by hh** [comment]
@zz Go ahead and specify the command as your startup script.

File: config/jobs/kubernetes/sig-release/conformance-gate.yaml
Comment (feiskyer) [inline config/jobs/kubernetes/sig-release/conformance-gate.yaml:8]: could we avoid using a personal image here?

**Comment by spiffxp** [comment]
>You can see an example of this job on prow.cncf.io:
> - the logs from a passing job( as 1.20.0 currently has no new untested endpoints)
> - the logs from a purposefully failing job(it looks at untested endpoints from prior releases)

Do you have a spyglass or gcsweb uri for those?  They expired

File: config/jobs/kubernetes/sig-release/conformance-gate.yaml
Comment (BenTheElder) [inline config/jobs/kubernetes/sig-release/conformance-gate.yaml:13]: FWIW there are a number of standard environment variables present in kubernetes CI
https://github.com/kubernetes/test-infra/blob/master/prow/jobs.md#job-environment-variables

**Comment by aojea** [comment]
2 questions:

1. can we have another job to gate per percentage? i.e. fail if GA less than 75%, beta less than 50%
2. can we use a different terminology to differentiate between API endpoints and the endpoints API?

File: config/jobs/kubernetes/sig-release/conformance-gate.yaml
Comment (zachmandeville) [reply config/jobs/kubernetes/sig-release/conformance-gate.yaml:8]: @feiskyer , yep!  We just put in a change to APISnoop so that the new database it uses can be a part of the existing test-infra image builds for APISnoop.  I opened a PR for this change here: https://github.com/kubernetes/test-infra/pull/19234

When merged, I can update the container here to use this image.

**Comment by zachmandeville** [comment]
> > You can see an example of this job on prow.cncf.io:
> > 
> > * the logs from a passing job( as 1.20.0 currently has no new untested endpoints)
> > * the logs from a purposefully failing job(it looks at untested endpoints from prior releases)
> 
> Do you have a spyglass or gcsweb uri for those? They expired

@spiffxp , I do!  The logs are coming from jobs running at https://prow.cncf.io
Here's a link to [the logs for the most recent job](https://prow.cncf.io/log?job=snoop-gate&id=1306080617657536512)

(note: this job is now failing as 1.20.0 has a new, untested stable endpoint with `getInternalApiserverAPIGroup`)

File: config/jobs/kubernetes/sig-release/conformance-gate.yaml
Comment (zachmandeville) [reply config/jobs/kubernetes/sig-release/conformance-gate.yaml]: Good point, @BenTheElder!  I updated the underlying image to use a standard env var, and removed my redundant one.  This is part of commit a0c307c431521046333762492c402edab4e00c61

**Comment by zachmandeville** [comment]
> 2 questions:
> 
>     1. can we have another job to gate per percentage? i.e. fail if GA less than 75%, beta less than 50%
> 
>     2. can we use a different terminology to differentiate between API endpoints and the endpoints API?
 
To answer your questions, @aojea:
1. This would be straightforward to do with the foundations set in this job.
2. It would be straightforward to update the messages given in the logs to use different terminology, if this is what you mean?

File: config/jobs/kubernetes/sig-release/conformance-gate.yaml
Comment (zachmandeville) [reply config/jobs/kubernetes/sig-release/conformance-gate.yaml]: @feiskyer , This job now uses an image from `gcr.io/k8s-staging-apisnoop`.

**Comment by hh** [comment]
It's owned by SIG-Arch, specifically the Conformance Subproject, but I the folks that need to be notified are SIG-Release.

**Comment by dims** [comment]
Confirmed what @hh says

**Comment by justaugustus** [comment]
Okay, so we should probably have this under a SIG Arch directory, so we can assign conformance OWNERS to review/approve.

This was about to merge as a SIG Release test w/o SIG Release approval.

**Comment by zachmandeville** [comment]
I created a `sig-arch` directory at the same level as `sig-release` and moved this job there.  Would this work?
I also added an email for job fail notifications, but am keen for feedback on the best email to use for this!

**Comment by hh** [comment]
I'm trying to make sure we include @kubernetes/sig-release in the configuration of this job, but in crossing SIGs with regards to who's the consumer of this job vs who 'OWNERS' owns the reconfiguration via git.

The intent of the job is to be sure that beta to stable API promotions include conformance test coverage. This should be a release blocking job, so we don't create any more technical/testing debt.

The notification target / email should probably be sig-release rather than sig-arch, while the maintenance / ownership of the job itself should probably be sig-arch.

If this job fails, sig-release should communicate to the SIG/KEP/API introducing any newly stable feature their requirement to making this job pass by writing tests (which in turn get approved by SIG-Arch) within the release.

If the tests can't be written and promoted to satisfy sig-arch, then the new endpoints should be reverted to delayed to a later release.

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (neolit123) [inline config/jobs/kubernetes/sig-arch/conformance-gate.yaml:7]: This should be the sig arch mailing list?

**Review by neolit123 — COMMENTED** [review]
Please keep the commits sqashed to 1.

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (neolit123) [reply config/jobs/kubernetes/sig-arch/conformance-gate.yaml:7]: There should be a description annotation explaining with a couple of sentences what the job does. Example https://github.com/kubernetes/test-infra/blob/b660a5251071db1e5aca5a9e1ca5b08064691eb7/config/jobs/kubernetes/sig-node/node-kubelet.yaml#L5

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (spiffxp) [inline config/jobs/kubernetes/sig-arch/conformance-gate.yaml:16]: Can this be dropped?

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (spiffxp) [reply config/jobs/kubernetes/sig-arch/conformance-gate.yaml:7]: Agree with @neolit123's comments, although I might hold off on alerting the arch mailing list until this is proven to run and not fail spuriously.  The existing e-mail address is not valid though, I'd maybe suggest apisnoop@ii.coop to start with?

If the goal is to say sig arch owns this job, it would be helpful to setup a sig-arch dashboard group, eg: sig-arch/sig-arch-conformance/apisnoop-conformance-gate or something similar.

**Review by spiffxp — CHANGES_REQUESTED** [review]
Some comments/questions

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (zachmandeville) [reply config/jobs/kubernetes/sig-arch/conformance-gate.yaml]: ah, yep!  This has been dropped, @spiffxp.

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (zachmandeville) [reply config/jobs/kubernetes/sig-arch/conformance-gate.yaml]: @spiffxp and @neolit123 : I've included the creation of a sig-arch dashboard into this pr and updated the job's annotations accordingly.  The dashboard group is `sig-arch`, and the dashboard-name is `sig-arch-conformance`.

**Comment by spiffxp** [comment]
/cc @johnbelamaric
LGTM, but I'd like to get the /lgtm from @johnbelamaric as sig-arch chair

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (johnbelamaric) [inline config/jobs/kubernetes/sig-arch/conformance-gate.yaml:7]: Change to newly created group: kubernetes-sig-arch-conformance-test-failures@googlegroups.com

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
LGTM in general, please update to use the test failures group alias I just created.

File: config/jobs/kubernetes/sig-arch/conformance-gate.yaml
Comment (zachmandeville) [reply config/jobs/kubernetes/sig-arch/conformance-gate.yaml]: Ah, thank you for that!  The email has been updated.

**Comment by harche** [comment]
sorry my bad. Wrong repo.

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
